### ISSUE TITLE
Initial build is faster, but incremental builds are much more slower

### ISSUE THREAD

**Issue by zuzusik** [TypeStrong/fork-ts-checker-webpack-plugin#65]
So, my initial build time is now `30..40` seconds VS `60..80`, but in incremental builds only type checking takes `4..6` seconds VS  `1.5...2` for the whole build which I had before - this is a major degradation.

My config is:

```js
       {
          test: /\.ts$/,
          use: [
            {
              loader: 'ng-annotate-loader',
              options: {
                map: false,
                add: true
              }
            },
            {loader: 'cache-loader'},
            {
              loader: 'thread-loader',
              options: {
                // there should be 1 cpu for the fork-ts-checker-webpack-plugin
                workers: require('os').cpus().length - 1
              }
            },
            {
              loader: 'ts-loader',
              options: {happyPackMode: true}
            }
          ]
        },
// ...
plugins: [
  new ForkTsCheckerWebpackPlugin({checkSyntacticErrors: true})
],
```
my cpu as reported by Node is 
```js
model: 'Intel(R) Core(TM) i7-6567U CPU @ 3.30GHz',   
speed: 3300,
```
I am running it on MacBook Pro 2016, 13" with 16GB of RAM
Node version - `6.9.5`
OS version - 10.12.6 / 10.13.1 (tested on both)

Things I've tried to do to fix a problem and they didn't help:
- specifying 2 workers for `ForkTsCheckerWebpackPlugin`
- removing `cache-loader`
- removing `thread-loader`
- removing both `cache-` and `thread-` loaders
- removing `ng-annotate-loader`
- using `watch` config property
- restarting the system
- reinstalling npm dependencies from the scratch
- upgrading OS from latest Sierra (10.12.6) to latest High Sierra (10.13.1)

previous webpack config was:

```js
       {
          test: /\.ts$/,
          use: [
            {
              loader: 'ng-annotate-loader',
              options: {
                map: false,
                add: true
              }
            },
            {loader: 'ts-loader'}
          ]
        },
```
I also noticed that when I remove `thread-loader` webpack generates `.js` file (this triggers `gulp-livereload` task in our case) before type checking is finished, but if I enable `thread-loader` then `.js` file is only generated after type-checking is finished.

**Comment by ivanvoznyakovsky** [comment]
Same problem on MBP late 2013, 16GB RAM and Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz. Incremental build slow down from `1.2...1.5` to `2.5...2.8` seconds.

**Comment by johnnyreilly** [comment]
Maybe worth fiddling with your settings?

**Comment by zuzusik** [comment]
@johnnyreilly if you are talking about @ivanvoznyakovsky's settings - they are the same as mine, we are working on the same project

if you need more settings provided by me, please specify what you need - will be happy to provide it

**Comment by johnnyreilly** [comment]
I should say that fork-ts-checker-webpack-plugin is not my project.  That said, I'd suggest moving thread-loader and cache-loader out of the mix for now.  Focus just on using the plugin.  I can see you've tried that; I'm surprised it didn't make any difference.

I can't debug your problem for you but I'd advise taking a look at some working setups that you  might find here:  https://github.com/TypeStrong/ts-loader/tree/master/examples

Both this: https://github.com/TypeStrong/ts-loader/tree/master/examples/fork-ts-checker
and this: https://github.com/TypeStrong/ts-loader/tree/master/examples/core-js

are probably worth a gander.

**Comment by zuzusik** [comment]
> I can't debug your problem for you 

Maybe you can advice the way we should try to debug it? Maybe some additional diagnostic information would be helpful for you?

> I can see you've tried that; I'm surprised it didn't make any difference.

it makes *some* difference - it makes things like 10% slower :)

> I'd advise taking a look at some working setups

where is the guarantee that in that setups the incremental build speed is not slowed down?

I can see the docs mentioning `~3s` rebuild time which sounds too big for me (we have really big project and as you see, it was 1.5..2 seconds on slower CPU and 1.2..1.5 seconds on faster one) 

Have you tried to measure rebuild time without this plugin? Is it faster/slower? Maybe it is a general thing for this plugin to speed up initial build and slow down incremental builds?

**Comment by johnnyreilly** [comment]
I'm really sorry - I'd love to help more but right now I am snowed.

**Comment by zuzusik** [comment]
@johnnyreilly sure 👍 just trying to understand whether this is just our project issue or the general issue of this plugin

**Comment by johnnyreilly** [comment]
I'd suggest this is something to do with your project. The only purpose of this plugin is to speed up builds, with particular focus on incremental build performance. Probably the most useful things you could do is create a minimal reproduction of your slow build and share it on GitHub. That will allow someone to look at your setup and advise. Speaking from my own experience, it's notable that during the creation of a minimal repo, the solution can present itself unexpectedly :wink:

**Comment by zuzusik** [comment]
@johnnyreilly the thing is that our repo is not **minimal** - it has hundreds of our own files and tens of external npm modules imported all over - so I can suppose this incremental builds slowing down issue can stand out only on big projects

So, my question would be - was this tool tested on really big projects? If yes - where I can look at such tests/real life usage examples? If no - maybe you can give some advices how to get such a project? Maybe some other repos already have samples of **big** TS projects?

**Comment by piotr-oles** [comment]
I've created this project because we had very slow incremental build time (about 40 seconds). The project in this stage had about 100k lines of code. Thanks to this plugin our incremental check dropped to ~10 seconds and didn't block emit. 

The problem with a semantic check is that it can't be done only for modified files - we need to perform this check on all project files (in fact we could limit it by dependency graph, but there was no easy way using typescript API). Maybe we will be able to optimize this with some new typescript API.

PS. Older versions of TypeScript are faster because it performs fewer checks.
